src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/innerassignment/InputInnerAssignmentSwitchAndSwitchExpression.java

```diff
@@ -0,0 +1,85 @@
+/*
+InnerAssignment
+
+
+*/
+package com.puppycrawl.tools.checkstyle.checks.coding.innerassignment;
+
+//non-compiled with javac: Compilable with Java14
+public class InputInnerAssignmentSwitchAndSwitchExpression {
+
+    public void test1(int mode) {
+        int x = 0;
+        switch(mode) {
+            case 2 -> {
+                x = 2;
+            }
+            case 1 -> x = 1;
+        }
+    }
+
+    public void test2(int mode) {
+        int x = 0;
+        switch(mode) {
+            case 2, 4, 6 -> {
+                x = 2;
+            }
+            case 1, 3, 5 -> x = 1;
+        }
+    }
+
+    public void test3(int mode) {
+        int x = 0;
+        x = switch (mode) {
+            case 2 -> {
+                yield x = 2; // violation
```

Comment:
So why does this create violations but something like test 2 doesn't? Is this because it is basically nested inside an assignment already?

Is there somewhere (official?) that shows what inner assignments are in this case?

What if we had a case like (also add it):
````
System.out.println(switch (mode) {
            case 2 -> {
                yield x = 2;
````

This wouldn't be an exception?

Reply:
> So why does this create violations but something like test 2 doesn't

because this is an assignment inside expressions ( switch expressions ), while test2 is an assignment inside normal switch statements

> Is there somewhere (official?)

I didn't find any but what I know from our docs is that inner assignments are the assignments inside expressions. So basically the assignments inside switch expressions `x = switch(){....} `  should be avoided

> What if we had a case like (also add it):

Yes, violation and added

Reply:
>Is there somewhere (official?) that shows what inner assignments are in this case?

Doubt it. A better way to describe what this check does is "place violations (mostly) anywhere that assigned values are used directly". Ideally, we should only do one thing at a time; assign a value, or use it, not both.

Reply:
> yield x = 2;

Just to confirm, since I am not use to this syntax, is this the same as `x = 2; yield x;`?
If we had `y = 2` before this line would that be a violation too?

@nrmancuso @romani Are we ok that test8 is a violation for this check?

Reply:
> Just to confirm, since I am not use to this syntax, is this the same as `x = 2; yield x;`?

Yes I think So

> If we had `y = 2` before this line would that be a violation too?

this will have no violation due to https://github.com/checkstyle/checkstyle/blob/405db25d46c4cb9ac561f91e04c59aebac50baf6/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/InnerAssignmentCheck.java#L94

Reply:
Yes, assigned values are used directly in all three cases.